pkg/runtime/bulk_subscriber.go

```diff
@@ -0,0 +1,716 @@
+			}
+		}
+		var overallInvokeErr error
+		for path, psm := range routePathBulkMessageMap {
+			invokeErr := a.createEnvelopeAndInvokeSubscriber(ctx, psm, topic, psName, msg, route, &bulkResponses, &entryIdIndexMap, path, policy, &bulkSubDiag)
+			if invokeErr != nil {
+				hasAnyError = true
+				err = invokeErr
+				overallInvokeErr = invokeErr
+			}
+		}
+		if errors.Is(overallInvokeErr, context.Canceled) {
+			reportBulkSubDiagnostics(ctx, topic, &bulkSubDiag)
+			return bulkResponses, overallInvokeErr
+		}
+		if hasAnyError {
+			// Sending msg to dead letter queue.
+			// If no DLQ is configured, return error for backwards compatibility (component-level retry).
+			bulkSubDiag.retryReported = true
+			if dlqErr := a.sendBulkToDLQIfConfigured(ctx, psName, msg, route, &entryIdIndexMap, &bulkResponses, &bulkSubDiag); dlqErr != nil {
+				reportBulkSubDiagnostics(ctx, topic, &bulkSubDiag)
+				return bulkResponses, err
+			}
+			reportBulkSubDiagnostics(ctx, topic, &bulkSubDiag)
+			return nil, nil
+		}
+		reportBulkSubDiagnostics(ctx, topic, &bulkSubDiag)
+		return bulkResponses, err
+	}
+
+	if bulkSubscriber, ok := ps.component.(pubsub.BulkSubscriber); ok {
+		return bulkSubscriber.BulkSubscribe(ctx, req, bulkHandler)
+	}
+
+	return runtimePubsub.NewDefaultBulkSubscriber(ps.component).BulkSubscribe(ctx, req, bulkHandler)
+}
+
+// sendBulkToDLQIfConfigured sends the message to the dead letter queue if configured.
+func (a *DaprRuntime) sendBulkToDLQIfConfigured(ctx context.Context, psName string, msg *pubsub.BulkMessage, route TopicRouteElem,
+	entryIdIndexMap *map[string]int, bulkResponses *[]pubsub.BulkSubscribeResponseEntry, bulkSubDiag *bulkSubIngressDiagnostics, //nolint:stylecheck
+) error {
+	if route.deadLetterTopic != "" {
+		if dlqErr := a.sendBulkToDeadLetter(ctx, psName, msg, route.deadLetterTopic, nil, nil, bulkSubDiag); dlqErr == nil {
+			// dlq has been configured and whole bulk of messages is successfully sent to dlq.
+			return nil
+		}
+	}
+	if !bulkSubDiag.retryReported {
+		bulkSubDiag.statusWiseDiag[string(pubsub.Retry)] += int64(len(msg.Entries))
+	}
+	return errors.New("failed to send to DLQ as DLQ was not configured")
+}
+
+// getRouteIfProcessable returns the route path if the message is processable.
+func (a *DaprRuntime) getRouteIfProcessable(ctx context.Context, route TopicRouteElem, message *pubsub.BulkMessageEntry,
+	i int, bulkResponses *[]pubsub.BulkSubscribeResponseEntry, matchElem interface{},
+	psName string, topic string, bulkSubDiag *bulkSubIngressDiagnostics,
+) (string, error) {
+	rPath, shouldProcess, routeErr := findMatchingRoute(route.rules, matchElem)
+	if routeErr != nil {
+		log.Errorf("error finding matching route for event in bulk subscribe %s and topic %s for entry id %s: %s", psName, topic, message.EntryId, routeErr)
+		(*bulkResponses)[i].EntryId = message.EntryId
```

Comment:
Nit: All of these are a bit awkward. Could you create a variable that is already of the right type so this doesn't need to be addressed in each instance?

Reply:
Addressed with https://github.com/dapr/dapr/pull/5398